lib/Dialect/TritonGPU/Transforms/Pipeliner/LowerLoops.cpp

```diff
@@ -469,6 +469,28 @@ scf::ForOp lowerLoads(scf::ForOp forOp, CoarseSchedule &schedule,
           // Allocate additional buffer required by the wgmma pipelining.
           stageDiff += 1;
         }
+
+        // Check if we need extra buffer due to unusual execution order
+        // The issue occurs when users of the load are scheduled in a later
+        // cluster, which happens when conditional code gets moved to epilogue
+        // cluster. This creates a race condition where the local load happens
+        // after the global-to-local copy for the next pipeline stage starts.
+        // Identify the cluster containing the load
+        auto loadCluster = schedule[&op].second;
+        // Check if any users are scheduled in a later cluster
+        for (Operation *user : triton::getTopLevelUsersInLoop(&op, forOp)) {
+          if (schedule.count(user)) {
+            auto userCluster = schedule[user].second;
+            // If the user is in a later cluster than the load, the local load
+            // will happen after the global-to-local copy starts, requiring
+            // extra buffer
+            if (*userCluster > *loadCluster) {
+              stageDiff += 1;
+              break;
+            }
+          }
+        }
```

Comment:
ouch, yes this is a major miss. We need to bump the number of buffers only if the user is in the last stage (there could be multiple users in different stages). I wonder if we should bake this logic in `getDefUseStageDiff` instead.
@pawelszczerbuk do you have an opinion?

Reply:
copy stage 1 cluster 1 -> use stage 2 cluster 2. This needs 2 buffers, even though stageDiff is 1. This PR would provide that increment from 1 to 2.

copy stage 1 cluster 1 -> use stage 3 cluster 0. This needs 2 buffers, which is already provided because stageDiff is 2.

So you're saying that with my PR, `getDefUseStageDiff` would return a stageDiff of 2 because of the latter dependency, then it would detect the former dependency and increment, incorrectly using 3 buffers total.

I think you're correct, and your suggestion to put the logic instead in `getDefUseStageDiff` makes sense, I'll try and do that. 👍